From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nobody <nobody@nobody>
Date: Mon, 22 Oct 2018 10:45:53 -0700
Subject: ignore_mouse_buttons_other_than_left_right_middle.patch

disable navigation via mouse4 and mouse5

diff --git a/ui/views/win/hwnd_message_handler.cc b/ui/views/win/hwnd_message_handler.cc
index 037ace982ed86c515e7b024d0cd2862d604ec63b..882b4264c8fa91cc4f46e3b0c7ca49dd31e43954 100644
--- a/ui/views/win/hwnd_message_handler.cc
+++ b/ui/views/win/hwnd_message_handler.cc
@@ -1079,6 +1079,11 @@ LRESULT HWNDMessageHandler::HandleMouseMessage(unsigned int message,
                                                WPARAM w_param,
                                                LPARAM l_param,
                                                bool* handled) {
+  if (message == WM_XBUTTONUP || message == WM_XBUTTONDOWN || message == WM_XBUTTONDBLCLK) {
+    *handled = false;
+    return 0;
+  }
+
   // Don't track forwarded mouse messages. We expect the caller to track the
   // mouse.
   base::WeakPtr<HWNDMessageHandler> ref(msg_handler_weak_factory_.GetWeakPtr());
